interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp

```diff
@@ -59,26 +59,128 @@ namespace cling {
       addSearchPath(P, /*IsUser*/ false);
   }
 
+  ///\returns substitution of pattern in the front of original with replacement
+  /// Example: substFront("@rpath/abc", "@rpath/", "/tmp") -> "/tmp/abc"
+  static std::string substFront(llvm::StringRef original, std::string pattern,
+                                llvm::StringRef replacement) {
+    if (!original.startswith_lower(pattern))
+      return original.str();
+    llvm::SmallString<512> result(replacement);
+    result.append(original.drop_front(pattern.length()).str());
+    return result.str();
+  }
+
+  ///\returns substitution of all known linker variables in \c original
+  static std::string substAll(llvm::StringRef original,
+                              llvm::StringRef libLoader) {
+
+    // Handle substitutions (MacOS):
+    // @rpath - This function does not substitute @rpath, becouse
+    //          this variable is already handled by lookupLibrary where
+    //          @rpath is replaced with all paths from RPATH one by one.
+    // @executable_path - Main program path.
+    // @loader_path - Loader library (or main program) path.
+    //
+    // Handle substitutions (Linux):
+    // https://man7.org/linux/man-pages/man8/ld.so.8.html
+    // $origin - Loader library (or main program) path.
+    // $lib - lib lib64
+    // $platform - x86_64 AT_PLATFORM
+
+    std::string result;
+    llvm::SmallString<512> mainExecutablePath(llvm::sys::fs::getMainExecutable(nullptr, nullptr));
+    llvm::sys::path::remove_filename(mainExecutablePath);
+    llvm::SmallString<512> loaderPath;
+    if (libLoader.empty()) {
+      loaderPath = mainExecutablePath;
+    } else {
+      loaderPath = libLoader.str();
+      llvm::sys::path::remove_filename(loaderPath);
+    }
```

Comment:
Isn't this equivalent and less CPU and simpler logic?
```suggestion
    llvm::SmallString<512> loaderPath;
    if (libLoader.empty()) {
      loaderPath = llvm::sys::fs::getMainExecutable(nullptr, nullptr);
    } else {
      loaderPath = libLoader.str();
    }
    llvm::sys::path::remove_filename(loaderPath);
```

Reply:
We use the `mainExecutablePath` below.